torch/nn/modules/transformer.py

```diff
@@ -0,0 +1,290 @@
+        dim_feedforward: the dimension of the feedforward network model (default=2048).
+        dropout: the dropout value (default=0.1).
+
+    Examples::
+        >>> transformer_model = nn.Transformer(src_vocab, tgt_vocab)
+        >>> transformer_model = nn.Transformer(src_vocab, tgt_vocab, nhead=16, num_encoder_layers=12)
+    """
+
+    def __init__(self, d_model=512, nhead=8, num_encoder_layers=6,
+                 num_decoder_layers=6, dim_feedforward=2048, dropout=0.1):
+        encoder_layer = TransformerEncoderLayer(d_model, nhead, dim_feedforward, dropout)
+        encoder_norm = LayerNorm(d_model) 
+        encoder = TransformerEncoder(encoder_layer, num_encoder_layers, encoder_norm)
+
+        decoder_layer = TransformerDecoderLayer(d_model, nhead, dim_feedforward, dropout)    
+        decoder_norm = LayerNorm(d_model) 
+        decoder = TransformerDecoder(decoder_layer, num_decoder_layers, decoder_norm)
+
+        super(Transformer, self).__init__(encoder, decoder)
+
+        self._reset_parameters()
+
+    def forward(self, src, tgt, src_mask=None, tgt_mask=None, memory_mask=None):
+        r"""Take in and process masked source/target sequences.
+
+        Args:
+            src: the sequence to the encoder (required).
+            tgt: the sequence to the decoder (required).
+            src_mask: the mask for the src sequence (optional).
+            tgt_mask: the mask for the tgt sequence (optional).
+            memory_mask: the mask for the encoder output (optional).
+
+        Shape:
+            src: :math:`(S, N, E)`
+            tgt: :math:`(T, N, E)`
+            src_mask: math:`(S, S)`
+            tgt_mask: :math:`(T, T)`
+            memory_mask: :math:`(T, S)`
+            Note: The maksed positions are filled with float('-inf'). 
+                  Unmasked positions are filled with float(0.0). Masks ensure that the predictions 
+                  for position i depend only on the information before position i.
+
+            output: :math:`(T, N, E)` 
+            Note: Due to the multi-head attention architecture in the transformer model, 
+                  the output sequence length of a transformer is same as the input sequence
+                  (i.e. target) length of the decode. 
+
+            where S is the source sequence length, T is the target sequence length,
+            N is the batch size, E is the feature number
+
+        Examples:
+            >>> output = transformer_model(src, tgt, src_mask=src_mask, tgt_mask=tgt_mask)
+        """
+
+        memory = self.encoder(src, src_mask)
+        output = self.decoder(tgt, memory, tgt_mask, memory_mask)
+        return output
+
+    def generate_square_subsequent_mask(self, sz):
+        r"""Generate a square mask for the sequence. The masked positions are filled with float('-inf').
+            Unmasked positions are filled with float(0.0).
+        """
+        attn_shape = (sz, sz)
+        mask = np.triu(np.ones(attn_shape), k=1).astype('uint8')
```

Comment:
we don't have hard dependency on numpy

Reply:
this function also seems inefficient and unused everywhere

Reply:
Will fix it.